Saber/DepthBuffer.h

```diff
@@ -43,15 +50,109 @@ class DepthBuffer {
 		UINT64 width,
 		UINT height
 	);
+	void ResizeHZB(
+		Microsoft::WRL::ComPtr<ID3D12Device2> pDevice,
+		Microsoft::WRL::ComPtr<D3D12MA::Allocator> pAllocator,
+		UINT64 width,
+		UINT height
+	);
 
 	void Clear(Microsoft::WRL::ComPtr<ID3D12GraphicsCommandList2> pCommandList);
 
+	void SetSinglePassDownsampler(
+		std::shared_ptr<SinglePassDownsampler> pSPD,
+		Microsoft::WRL::ComPtr<ID3D12Device2> pDevice,
+		Microsoft::WRL::ComPtr<D3D12MA::Allocator> pAllocator,
+		UINT64 width,
+		UINT height
+	);
+
+	void CreateHierarchicalDepthBuffer(
+		Microsoft::WRL::ComPtr<ID3D12GraphicsCommandList2> pCommandList,
+		Microsoft::WRL::ComPtr<ID3D12DescriptorHeap> pDescHeap
+	) {
+		if (!m_pSinglePassDownsampler) {
+			return;
+		}
+		//ResourceTransition(pCommandList, m_pDepthBuffer->GetResource(),
+		//	D3D12_RESOURCE_STATE_DEPTH_WRITE,
+		//	D3D12_RESOURCE_STATE_ALL_SHADER_RESOURCE
+		//);
+		//ResourceTransition(pCommandList, m_pHZBuffer->GetResource(),
+		//	D3D12_RESOURCE_STATE_ALL_SHADER_RESOURCE,
+		//	D3D12_RESOURCE_STATE_UNORDERED_ACCESS
+		//);
+		//ResourceTransition(pCommandList, m_pDepthBuffer->GetResource(),
+		//	D3D12_RESOURCE_STATE_DEPTH_WRITE,
+		//	D3D12_RESOURCE_STATE_COPY_SOURCE
+		//);
+		//ResourceTransition(pCommandList, m_pHZBuffer->GetResource(),
+		//	D3D12_RESOURCE_STATE_ALL_SHADER_RESOURCE,
+		//	D3D12_RESOURCE_STATE_COPY_DEST
+		//);
+		//pCommandList->CopyResource(
+		//	m_pHZBuffer->GetResource().Get(),
+		//	m_pDepthBuffer->GetResource().Get()
+		//);
+		//ResourceTransition(pCommandList, m_pDepthBuffer->GetResource(),
+		//	D3D12_RESOURCE_STATE_COPY_SOURCE,
+		//	D3D12_RESOURCE_STATE_DEPTH_READ
+		//);
+		//ResourceTransition(pCommandList, m_pHZBuffer->GetResource(),
+		//	D3D12_RESOURCE_STATE_COPY_DEST,
+		//	D3D12_RESOURCE_STATE_UNORDERED_ACCESS
+		//);
+		//pCommandList->CopyTextureRegion(
+		//	&CD3DX12_TEXTURE_COPY_LOCATION(m_pHZBuffer->GetResource().Get(), 0),
+		//	0, 0, 0,
+		//	&CD3DX12_TEXTURE_COPY_LOCATION(m_pDepthBuffer->GetResource().Get(), 0),
+		//	&CD3DX12_BOX(0, 0, 0, m_width, m_height, 1)
+		//);
+		
+		pCommandList->CopyTextureRegion(
```

Comment:
copy will be on compute queue?

Reply:
Everything on direct queue during rendering